### ISSUE TITLE
Add `map.setSprite(spriteUrl)`method

### ISSUE THREAD

**Issue by scothis** [mapbox/mapbox-gl-js#2058]
The style diff utility defines an [operation to change the sprite sheet](https://github.com/mapbox/mapbox-gl-style-spec/blob/ef23dd1e8141a60bac2a0ed0fe3ec2ec9e792d82/lib/diff.js#L77-L80) for a style. Mapbox GL JS does not support this method currently causing a fall back to the more invasive setStyle when this operation is encountered.

When this method is invoked, the current sprite sheet should be removed and recreated.

Minimally `setSprite(spriteUrl)` should be added to the style batch, it would also be helpful to expose this method on the map instance.

**Comment by jfirebaugh** [comment]
A bit tricky to implement, but this should be supported. (As should changing the glyphs URL.)

The implementation will need to assume that the new sprite could have added icons, removed icons, and changed the contents, sizes, and pixel ratios of existing icons. All tiles will need to be re-layouted, and in order to avoid flickering and misrendering, the existing sprite texture will need to be preserved until all tiles have completed relayout. Basically, multiple sprite atlases will need to exist in parallel, and a tile will need to indicate which atlas it was layouted against.

**Comment by yuvadm** [comment]
This would also help #2081 since currently it seems the `sprite` attribute has no effect

**Comment by lucaswoj** [comment]
Closing in favor of mapbox/mapbox-gl-js#2059 & mapbox/mapbox-gl-style-spec#220

**Comment by anandthakker** [comment]
Reopening this as a shorter-term fix that is required before https://github.com/mapbox/mapbox-gl-js/pull/3643 can merge.

**Comment by anandthakker** [comment]
Now that we've decided this doesn't block #3643, I'm going to re-close this issue in favor of mapbox/mapbox-gl-js#2059

**Comment by mourner** [comment]
Reopening because we're still interested in the original use case — speeding up `setStyle` in the case the spritesheet is changed.

**Comment by ggerber** [comment]
I have two custom geojson sources and layers that I add to my map.
I would like the user to change the 'baselayer' e.g from streets to satellite.
The first attempt was:
```
map.setStyle(`mapbox://styles/mapbox/${_id}`);
```
This changes the the 'baselayer' but does not render the custom layers. On the console I get this warning:
```
Unable to perform style diff: Unimplemented: setSprite..  Rebuilding the style from scratch.
```
The second attempt was to remove the custom layers and sources, then do map.setStyle, then add the custom layers back:
```
    map.removeLayer('customlayer');
    map.removeSource('customsource');
    map.setStyle(`mapbox://styles/mapbox/${_id}`);
    map.addSource(...);
    map.addLayer(....);
```
This gives the same result as the first attempt: changes the baselayer, but does not render the custom layers with the following console message:
```
Unable to perform style diff: Unimplemented: setSprite..  Rebuilding the style from scratch.
```
I am able to achieve my goal through the following brute force approach:
1. Get and save the current camera parameters (map.getBearing(), etc)
2. Calling map.remove()
3. Rebuilding a brand new map instance with the save camera parameters
4. Adding the custom sources and layers to the new map instance.

It appears to me to be quite drastic to tear down the whole map instance, simply to change the baselayer.

Are there better ways of achieving this goal or should we wait till setStyle()/setSprite() is improved in the future?

**Comment by andrewharvey** [comment]
@ggerber #4006 is probably closer to what you're after since you want to retain your application layers when changing the style. Those warnings you mentioned shouldn't affect functionality as far as I know, only how efficiently it can change the style and doing it with smooth transitions.
